### ISSUE TITLE
Overlay the Object Term or DEN with a selected Business Term during Expression

### ISSUE THREAD

**Issue by dubnemo** [OAGi/Score#1310]
Overlay the Object Term or DEN with a selected Business Term during Expression

**Comment by sfohn** [comment]
I think we have to be careful here.  My understanding of Business Term is that it is part of a Business Language, separate from an Interchange Language.  While the Business Language may be realized in the Interchange Language (e.g., OAGIS), there is no clear, deterministic relationship between terms in the two languages.  A given Interchange Language Term (by design to support internationalization requirements) may realize multiple Business Language Terms in a single context.  For example, the Interchange Term "GovernmentID" (with metadata to indicate scheme and schemeAgency) may support two Business Terms: ITIN (Individual Taxpayer Identification Number) and SSN (Social Security Number) in the US context. In this case, overlaying an Interchange Term with a Business Term is not possible.  If we need to overlay Interchange Terms with other terms, then those other terms must still be part of an Interchange Language ("apples to apples", as they say_  One might consider, those "other terms" to be "translations" to an Interchange Language Dialect. Then, when one "overlays", one effectively translates across equivalent concepts from a base language term to a language dialect term.

**Comment by dubnemo** [comment]
@sfohn It becomes unwieldy to create our own versions of Party when an industry has terms they want expressed in an API definition.  The Use of Party -> Location -> LocationArea is not working in Agriculture when the terms Farmer -> Farm -> Field are what they really need, and if OAGIS is to be relevant in Agriculture, we would have to completely rebuild all the underlying components, and reassociate them.  For example Field would be based on Location Area Base.  I would to to rebuild Location Base to Farm Base adding Field ASCCP, etc.    I think it would take several days to do that, whereas this would let me do the same in minutes.

**Comment by dubnemo** [comment]
In addition, I would be preserving the lineage from Field to Location Area through its association of the business term, which has business context as well.

**Comment by sfohn** [comment]
@dubnemo yes, I understand and agree with the requirement. I also agree with the general approach of using a target translation in an expression to address the requirement.  My point, is simply, that a "blanket approach" that replaces interchange language terms with business term translations in expression will not always viable (per my example, above). This should be considered when developing a solution that uses translations in an expression.

**Comment by dubnemo** [comment]
@kbserm 
Related to @elena-jelisic's sides, and your email.  I believe that you are referring to this specific issue.
I want to use this issue instead of email related to this topic.

Today we have the ability to create user defined core components, mostly based on industry context.  UDCCs are generally based on existing components and BDTs.  For example, in AgGateway, we created a Scale Ticket UDCCs components for AgGateway grain delivery and feed shipment needs.  I initially suggested AgGateway use ReceiveDelivery from the Express Pack, but we reject that due to naming.   Inside ReceiveDelivery there is a ReceiveDeliveryUnit –that is the same physical thing as the ShipmentUnit (semi-trailer, rail car, etc.), but the context is inbound (PO Receipt /deliveries) vs outbound shipments (Ship Confirm).  So in this specific case, the business process context was previously (and oddly) built into the DEN name, and that caused ‘concerns’ as people don’t normally call a Shipment Unit (ship unit) a Receive Delivery Unit.  It is a failing of our rigid NDR.

By overloading the DEN with the business term, we believe I could have made ReceiveDelivery work.  We instead decided to create our own UDCCs.  

Another contextualized example is related to Farms, Fields and what we call Crop zones.  Today, we have a Party -> Location -> Location Area – Location Subdivision Area hierarchy, where the latter two have geospatial geometry support.  Location only has Coordinate.  I would have much preferred to re-label the DENs Location Area to Field and Location Subdivision Area to Crop Zone or Part Field.  So therefore the choice was pretty simple to the AgGateway folks – we won’t use the OAGIS components at all and build something new completely outside the OAGIS standard.  Today that is ADAPT.  This was pre-SCORE during the GEFEG time period.  I instead recommended ArgoUML to model this out.

I think this latter case is probably the more realistic and harsh.  Instead of having interoperability problems with the standard, the choice was not to use the standard at all and build something from scratch.  That to me leads to much bigger interoperability problems than having a parent DEN re-labeled to a business term based on its context.

Last note: I believe interoperability is about what the components are based on, less the name of the top-level DEN.  No one will send a Scale Ticket when a ReceiveDelivery was expect.  That is why we have dev, QA and Prod environments for developers.  Similarly-based components are obviously very similar in structure.  During Code Gen, you will see that most components will have its own generated class, and the generated class will not differ much from the when same component is used in another components (WHERE USED).  That makes the developer happier due to recognition.

**Comment by elena-jelisic** [comment]
@dubnemo I completely agree with everything you said. That is exactly how I am looking at the interoperability problem. 

@sfohn Regarding Steffen's concerns (from the earlier discussion), I would say that the precise definition of a business context is crucial when we are talking about business terms. I don't know what other folks think, but I would restrict it to having only one business term per business context (per DEN). In that way, we wouldn't get into situations that you described, Steffen. Furthermore, yes, that is a valid example, BUT there is something else that dictates which term should be used and when. I would expect that business context is precise enough to "say" in which situations ITIN or SSN should be used. For example, for tax return purposes, a person will use the ITIN, if he does not have SSN (this should be a part of a business context - "he is not eligible to get one").

**Comment by sfohn** [comment]
Hi Elena @elena-jelisic.  The primary point that I was raising (which may or may not be an issue), is the intermingling (e.g., substitution) of the interchange language with business language (i.e., Business Terms).  It seems to me that an interchange language might well have a different "dialect" (e.g., an agriculture dialect, as Scott describes) that is more suited to particular domain or industry.  However, it doesn't seem that an industry "dialect" of the interchange is necessarily equivalent to the business language.  In my example of the interchange language term,  "GovernmentID", the related business terms, ITIN (Individual Taxpayer Identification Number) and SSN (Social Security Number) while associated to a US context, would not necessarily change the expression of interchange language; it is still beneficial to use "GovernmentID" as the interchange language term in a US context in an HR standard, considering internationalization requirements for both interchange standards and global systems.
